awscli/customizations/wizard/core.py

```diff
@@ -0,0 +1,237 @@
+        if 'next_step' not in step:
+            return
+        return self._decide_next_step_name(step['next_step'])
+
+    def _decide_next_step_name(self, next_step):
+        if isinstance(next_step, str):
+            # If an explicit step name is provided, we don't need to do
+            # anything further.
+            return next_step
+        if 'switch' in next_step:
+            varname = next_step['switch']
+            value = self._parameters.get(varname)
+            return next_step[value]
+
+
+class BaseStep(object):
+    def run_step(self, step_definition, parameters):
+        raise NotImplementedError("run_step")
+
+
+class StaticStep(BaseStep):
+    def run_step(self, step_definition, parameters):
+        return step_definition['value']
+
+
+class PromptStep(BaseStep):
+    def __init__(self, prompter):
+        self._prompter = prompter
+
+    def run_step(self, step_definition, parameters):
+        choices = self._get_choices(step_definition, parameters)
+        response = self._prompter.prompt(step_definition['description'],
+                                         choices=choices)
+        return response
+
+    def _get_choices(self, step_definition, parameters):
+        choices = step_definition.get('choices')
+        if choices is not None:
+            if isinstance(choices, str):
+                # If 'choices' is a string, we assume it's the name of
+                # a variable.
+                return parameters[choices]
+            return choices
+
+
+class TemplateStep(BaseStep):
+
+    def run_step(self, step_definition, parameters):
+        value = step_definition['value']
+        return value.format(**parameters)
+
+
+class APICallStep(BaseStep):
+    def __init__(self, api_invoker):
+        self._api_invoker = api_invoker
+
+    def run_step(self, step_definition, parameters):
+        service, op_name = step_definition['operation'].split('.', 1)
+        return self._api_invoker.invoke(
+            service=service, operation=op_name,
+            api_params=step_definition['params'],
+            plan_variables=parameters,
+            optional_api_params=step_definition.get('optional_params'),
+            query=step_definition.get('query'),
+        )
+
+
+class APIInvoker(object):
+    """This class contains shared logic for the apicall step.
+
+    The ``apicall`` in the planner and executor are slightly different
+    but share a lot of similar logic.  This class is used share logic
+    between the two steps.
+
+    """
+    def __init__(self, session):
+        self._session = session
+
+    def invoke(self, service, operation, api_params, plan_variables,
+               optional_api_params=None, query=None):
+        client = self._session.create_client(service)
+        resolved_params = self._resolve_params(
+            api_params, optional_api_params, plan_variables)
+        response = getattr(client, xform_name(operation))(**resolved_params)
```

Comment:
Might be worth adding a TODO here that we'll need to convert resolved parameters to the correct type, as everything read in from prompts is going to be a string. For now, this means we're limited to calling operations that only require string input. 

Might be worth exploring if typing in the wizard model or enforcing type via the service model is better (or maybe even both).

Reply:
Yeah good point.  I'll add a todo for now.  I think allowing type conversions in the wizard model would give the most control, but auto-converting based on the service model would be a nice convenience if you don't need special conversion behavior (which I would guess is the most common).